voltdbclient/client.go

```diff
@@ -0,0 +1,240 @@
+	client.writeLoginMessage(&login)
+	if client.connData, err = client.readLoginResponse(); err != nil {
+		return err
+	}
+	client.clientHandle = 0
+	client.netListener = NewListener(client.reader)
+	client.netListener.start()
+	return nil
+}
+
+// Close a client if open. A Client, once closed, has no further use.
+// To open a new client, use NewClient.
+func (client *Client) Close() error {
+	var err error = nil
+	if client.reader != nil {
+		tcpConn := client.reader.(*net.TCPConn)
+		tcpConn.Close()
+	}
+	client.reader = nil
+	client.writer = nil
+	client.connData = nil
+	return err
+}
+
+// GoString provides a default printable format for Client.
+func (client *Client) GoString() string {
+	if client.connData != nil {
+		return fmt.Sprintf("hostId:%v, connId:%v, leaderAddr:%v buildString:%v",
+			client.connData.hostId, client.connData.connId,
+			client.connData.leaderAddr, client.connData.buildString)
+	}
+	return "uninitialized"
+}
+
+// MultiplexCallbacks 'fans in' callbacks - listens for the given set of callbacks on one channel
+func (client *Client) MultiplexCallbacks(callbacks []*Callback) <-chan *Response {
+	c := make(chan *Response)
+	for _, callback := range callbacks {
+		ch := callback.Channel
+		go func() {
+			c <- <-ch
+		}()
+	}
+	return c
+}
+
+// Ping the database for liveness.
+func (client *Client) PingConnection() bool {
+	if client.writer == nil {
+		return false
+	}
+	rsp, err := client.Call("@Ping")
+	if err != nil {
+		return false
+	}
+	return rsp.Status() == SUCCESS
+}
+
+// functions private to this package.
+
+// readLoginResponse parses the login response message.
+func (client *Client) readLoginResponse() (*connectionData, error) {
+	buf, err := readMessage(client.reader)
+	if err != nil {
+		return nil, err
+	}
+	connData, err := deserializeLoginResponse(buf)
+	return connData, err
+}
+
+// writeLoginMessage writes a login message to the connection.
+func (client *Client) writeLoginMessage(buf *bytes.Buffer) {
+	// length includes protocol version.
+	length := buf.Len() + 2
+	var netmsg bytes.Buffer
+	writeInt(&netmsg, int32(length))
+	writeProtoVersion(&netmsg)
+	writePasswordHashVersion(&netmsg)
+	// 1 copy + 1 n/w write benchmarks faster than 2 n/w writes.
+	io.Copy(&netmsg, buf)
+	io.Copy(client.writer, &netmsg)
+}
+
+// writeProcedureCall serializes a procedure call and writes it to a tcp connection.
+func (client *Client) writeProcedureCall(writer io.Writer, procedure string, handle int64, params []interface{}) error {
+
+	var call bytes.Buffer
+	var err error
+
+	// Serialize the procedure call and its params.
+	// Use 0 for handle; it's not necessary in pure sync client.
+	if call, err = serializeCall(procedure, handle, params); err != nil {
+		return err
+	}
+
+	// todo: should prefer byte[] in all cases.
+	var netmsg bytes.Buffer
+	writeInt(&netmsg, int32(call.Len()))
+	io.Copy(&netmsg, &call)
+	io.Copy(writer, &netmsg)
+	return nil
+}
+
+func (nv *NullValue) ColType() int8 {
+	return nv.colType
+}
```

Comment:
who reads responses? I dont see invocation response messages read.
We should have a sync callprocedure method also which looks like one with SyncCallback style.